### PR TITLE
ChainerX + ChainerMN integraiton on MNIST

### PR DESCRIPTION
**Pull request by keisukefukuda**
This PR implements ChainerMN on ChainerX integration.

With this PR, the MNIST example works on ChainerX.

Note that you need to set the following environmental variables when you install Chainer fro m source to activate ChainerX.
```
export CHAINERX_BUILD_CUDA=1
export CHAINERX_ENABLE_BLAS=1
export CHAINER_BUILD_CHAINERX=1
```

```
export MAKEFLAGS=-j16
```
is also helpful.

### CODE REVIEW

File: examples/chainermn/mnist/train_mnist.py [hunk 1]
```diff
@@ -106,7 +114,10 @@ def main():
     # Some display and output extensions are necessary only for one worker.
     # (Otherwise, there would just be repeated outputs.)
     if comm.rank == 0:
-        trainer.extend(extensions.DumpGraph('main/loss'))
+        if device.xp is not chainerx:
+            # Disabled for ChainerX.
+            # See examples/mnist/train_mnist.py
```
Comment (belldandyxtq) [inline examples/chainermn/mnist/train_mnist.py]: Maybe we can figure out why it is temporally disabled in chainer example https://github.com/chainer/chainer/blob/edc929818b0021543d2b8dc773aa89b41ec06d09/examples/mnist/train_mnist.py#L97

File: chainermn/extensions/multi_node_evaluator.py [hunk 2]
```diff
@@ -26,6 +29,23 @@ def create_multi_node_evaluator(actual_evaluator, communicator):
...
+        # We need convert chainerx ndarray to Native array because
+        #   (1) allreduce_obj is used to compute global mean values
+        #   (2) allreduce_obj calls mpi4py.allreduce, which pickles the object
+        #   (3) chainerx.ndarray preserves CUDA device internally when pickled
+        #   (4) Error is caused when an ndarray is unpickled in another process
+        array0 = list(local_mean_dict.values())[0]
+        xp = chainer.backend.get_array_module(array0)
+        if xp == chx and array0.device.backend.name == 'cuda':
+            # Results of evaluation is fairly small, so
+            # the ndarray is transferred to CPU and allreduce()-ed.
```
Comment (belldandyxtq) [inline chainermn/extensions/multi_node_evaluator.py]: We can leave a comment here as there is a potential performance issue

File: chainermn/extensions/multi_node_evaluator.py [hunk 3]
```diff
@@ -26,6 +29,23 @@ def create_multi_node_evaluator(actual_evaluator, communicator):
 
     def new_evaluate(self):
         local_mean_dict = self._mn_original_evaluate()
+
+        # ChainerX support:
+        # We need convert chainerx ndarray to Native array because
+        #   (1) allreduce_obj is used to compute global mean values
+        #   (2) allreduce_obj calls mpi4py.allreduce, which pickles the object
+        #   (3) chainerx.ndarray preserves CUDA device internally when pickled
+        #   (4) Error is caused when an ndarray is unpickled in another process
```
Comment (belldandyxtq) [inline chainermn/extensions/multi_node_evaluator.py]: "An error will occur" ?

File: examples/chainermn/mnist/train_mnist.py (hunk 1 above)
Comment (keisukefukuda) [reply examples/chainermn/mnist/train_mnist.py]: I'm now consulting @niboshi -san.

File: chainermn/extensions/multi_node_evaluator.py (hunk 2 above)
Comment (keisukefukuda) [reply chainermn/extensions/multi_node_evaluator.py]: Fixed! Thanks

File: chainermn/extensions/multi_node_evaluator.py (hunk 3 above)
Comment (keisukefukuda) [reply chainermn/extensions/multi_node_evaluator.py]: Fixed! thanks

File: examples/chainermn/mnist/train_mnist.py (hunk 1 above)
Comment (niboshi) [reply examples/chainermn/mnist/train_mnist.py]: `DumpGraph` is disabled because ChainerX doesn't have a public API set to traverse computational graphs.

Comment (keisukefukuda) [reply examples/chainermn/mnist/train_mnist.py]: OK, understood! Thanks!

Comment (belldandyxtq) [reply examples/chainermn/mnist/train_mnist.py]: Can you please add a comment about this?

**Comment by keisukefukuda** [comment]
pfnci, test this please

File: examples/chainermn/mnist/train_mnist.py (hunk 1 above)
Comment (keisukefukuda) [reply examples/chainermn/mnist/train_mnist.py]: fixed!

**Comment by keisukefukuda** [comment]
pfnci, test this please.

**Comment by belldandyxtq** [comment]
LGTM. Unit tests will be added later #8031 
pfnci, test this please.

**Comment by belldandyxtq** [comment]
I am getting the following error in the test
```
Traceback (most recent call last):
  File "train_mnist.py", line 136, in <module>
    main()
  File "train_mnist.py", line 60, in main
    device = chainer.get_device('cuda:{}'.format(comm.intra_rank))
  File "tianqi/repository/chainer/chainer/backend.py", line 176, in get_device
    'specifiers.'.format(device_spec))
RuntimeError: Tried to parse ChainerX device specifier 'cuda:1', but ChainerX is not available. Note that device specifiers without '@' prefix are assumed to be ChainerX device specifiers.
Traceback (most recent call last):
  File "train_mnist.py", line 136, in <module>
    main()
  File "train_mnist.py", line 60, in main
    device = chainer.get_device('cuda:{}'.format(comm.intra_rank))
  File "tianqi/repository/chainer/chainer/backend.py", line 176, in get_device
    'specifiers.'.format(device_spec))
RuntimeError: Tried to parse ChainerX device specifier 'cuda:0', but ChainerX is not available. Note that device specifiers without '@' prefix are assumed to be ChainerX device specifiers.
```

**Comment by belldandyxtq** [comment]
This PR only covers the mnist, we will list up the rest of examples in a TODO

**Comment by belldandyxtq** [comment]
jenkins, test this please

**Comment by belldandyxtq** [comment]
jenkins, test this please

**Pull request merged**